src/algebra/module/zlattice.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2023 Xavier Roblot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Xavier Roblot
+-/
+import analysis.normed.order.basic
+import linear_algebra.finite_dimensional
+import measure_theory.group.fundamental_domain
+
+/-!
+# ℤ-lattices
+
+Let `E` be a finite dimensional vector space over a `normed_lattice_field` `K` that is also
+a `floor_ring`, e.g. `ℚ` or `ℝ`. A (full) ℤ-lattice `L` of `E` is a discrete subgroup of `E` such
+that `L` spans `E` over `K`.
+
+The ℤ-lattice `L` can be defined in two ways:
+* For `b` a basis of `E`, then `L : submodule.span ℤ (set.range b)` is a ℤ-lattice of `E`.
+* As `L : add_subgroup E` with the additional properties:
+  `(hd : ∀ r : ℝ, (L ∩ (metric.closed_ball 0 r)).finite)`, that is `L` is discrete
+  `(hs : submodule.span ℝ (L : set E) = ⊤)`, that is `L` spans `E` over `K`.
+
+## Main result
+* `zspan.is_add_fundamental_domain`: for `L : submodule.span ℤ (set.range b)` a ℤ-lattice, proves
+ that the set defined by `zspan.fundamental_domain` is indeed a fundamental domain.
+
+-/
+
+open_locale classical
+
+noncomputable theory
+
+namespace zspan
+
+open measure_theory measurable_set submodule
+
+variables {E ι : Type*}
+
+section normed_lattice_field
+
+variables {K : Type*} [normed_lattice_field K]
+variables [normed_add_comm_group E] [normed_space K E]
+variables (b : basis ι K E)
+
+/-- The fundamental domain of the ℤ-lattice spanned by `b`. See `zspan.is_add_fundamental_domain`
+for the proof that it is the fundamental domain. -/
+def fundamental_domain : set E := { m | ∀ i : ι, b.repr m i ∈ set.Ico (0 : K) 1 }
+
+variables [floor_ring K]
+
+section fintype
+
+variable [fintype ι]
+
+/-- The map that sends a vector of `E` to the element of the ℤ-lattice spanned by `b` obtained
+by rounding down its coordinates on the basis `b`. -/
+def floor : E → span ℤ (set.range b) :=
+λ m, finset.univ.sum (λ i, int.floor (b.repr m i) • b.restrict_scalars ℤ i)
+
+lemma floor_single (m : E) (i : ι) :
+  b.repr (floor b m) i = int.floor (b.repr m i) :=
```

Comment:
Why `single`?

Reply:
Yeah, it is not really a good name. What about `repr_floor_apply` and `repr_fract_apply'?

Reply:
Yeah, that looks better.